save-agent/src/nativeMain/kotlin/org/cqfn/save/agent/SaveAgent.kt

```diff
@@ -0,0 +1,155 @@
+     */
+    fun stop() {
+        println("Stopping agent")
+        httpClient.close()
+        if (saveProcessJob?.isActive == true) {
+            saveProcessJob?.cancel()
+        }
+        isStopped.getAndSet(true)
+    }
+
+    @Suppress("WHEN_WITHOUT_ELSE")  // when with sealed class
+    private suspend fun startHeartbeats() = coroutineScope {
+        println("Scheduling heartbeats")
+        while (isStopped.value.not()) {
+            val response = runCatching { sendHeartbeat() }
+            if (response.isSuccess) {
+                when (response.getOrNull()) {
+                    is NewJobResponse -> maybeStartSaveProcess()
+                    TerminatingResponse -> {
+                        println("Terminating the agent because termination signal has been received")
+                        stop()
+                    }
+                    EmptyResponse -> Unit  // do nothing
+                }
+            } else {
+                println("Exception during heartbeat: ${response.exceptionOrNull()?.message}")
+            }
+            // todo: start waiting after request was sent, not after response?
+            println("Waiting for ${config.heartbeatIntervalMillis} sec")
+            delay(config.heartbeatIntervalMillis)
+        }
+    }
+
+    private suspend fun maybeStartSaveProcess() = coroutineScope {
+        if (saveProcessJob?.isCompleted == false) {
+            println("Shouldn't start new process when there is the previous running")
+        } else {
+            saveProcessJob = launch {
+                // new job received from Orchestrator, spawning SAVE CLI process
+                startSaveProcess()
+            }
+        }
+    }
+
+    /**
+     * @return Unit
+     */
+    internal suspend fun startSaveProcess() = coroutineScope {
+        // blocking execution of OS process
+        // val code = saveAgent.runSave(emptyList())
+        state.value = AgentState.FINISHED
+        // todo: read data from files here
+        sendExecutionData(ExecutionData(emptyList()))
+        state.value = AgentState.IDLE
+    }
+
+    private fun runSave(cliArgs: List<String>) = platform.posix.system("./save ${cliArgs.joinToString(" ")}")
+
+    /**
+     * @return a [HeartbeatResponse] from Orchestrator
+     */
+    internal suspend fun sendHeartbeat(): HeartbeatResponse {
+        // log.trace("Sending heartbeat to $orchestratorUrl")
+        // if current state is IDLE or FINISHED, should accept new jobs as a response
+        return httpClient.post("${config.orchestratorUrl}/heartbeat") {
+            contentType(ContentType.Application.Json)
+            accept(ContentType.Application.Json)
+            body = Heartbeat(id, state.value, 0)
+        }
+    }
+
+    private suspend fun sendExecutionData(executionData: ExecutionData): Unit = coroutineScope {
+        var result = runCatching {
+            postExecutionData(executionData)
+        }
+        while (result.isFailure) {
```

Comment:
not a while(true), but should be limited to some number of attempts

Reply:
Do we then need a mechanism for orchestrator to initiate data sending? In case backend dies, 5 retries are not enough, but we don't want to lose data?